### ISSUE TITLE
configure script unable to check for C compiler

### ISSUE THREAD

**Issue by tn1997tn** [rmyorston/busybox-w32#297]
```
configure: error: no acceptable C compiler found in $PATH
See `config.log' for more details
```

It also fails to check for various utilities like sed,... even though they are all there!

**Comment by rmyorston** [comment]
I think we need more information about you're trying to do before we can help.

**Comment by tn1997tn** [comment]
> I think we need more information about you're trying to do before we can help.

Try it yourself: http://fox-toolkit.org/ftp/fox-snapshot.zip

**Comment by rmyorston** [comment]
A couple of problems with the `configure` script:
- The test for the path separator only works if `/bin/sh` exists.
- The tests for executables try adding extensions, but I couldn't find anything that set the variable which is supposed to list possible extensions.

I got `configure --build=x86_64-w64-mingw32` to work by setting these environment variables before running it:
```sh
export PATH_SEPARATOR=';'
export ac_executable_extensions='.exe'
```
Running `make` then got stuck almost immediately.

**Comment by tn1997tn** [comment]
> Running `make` then got stuck almost immediately.

Same experience, even though I pass the path to the utilities directly via environment variable like `SED="path/to/sed" CC="path/to/gcc" ./configure --build=x86_64-w64-mingw32`. The configure script is something generated by autoconf. I have no idea why it's this way.

**Comment by rmyorston** [comment]
The problem with the build getting stuck is due to this function in the `libtool` script (which is created from `ltmain.sh` during configuration):
```sh
# Convert file name or path ARG from MSYS format to w32 format.  Return
# result in func_convert_core_msys_to_w32_result.
func_convert_core_msys_to_w32 ()
{
  $debug_cmd

  # awkward: cmd appends spaces to result
  func_convert_core_msys_to_w32_result=`( cmd //c echo "$1" ) 2>/dev/null |
    $SED -e 's/[ ]*$//' -e "$sed_naive_backslashify"`
}
```
In the Cygwin world `/c` is considered to be a path referencing `C:\` so `cmd /c` has to be written as `cmd //c`. In the real world of Windows `//c` has no special meaning so `cmd` doesn't know it's supposed to be running a single command and enters interactive mode where it gets stuck.

I've no idea why libtool is using such a convoluted method to get `$1` into `sed`, but since we don't really need to "backslashify" it anyway the function can be reduced to:
```sh
func_convert_core_msys_to_w32 ()
{
  $debug_cmd

  func_convert_core_msys_to_w32_result="$1"
}
```
With that change the build gets much further. I don't know _how_ much further because I stopped it after a while.

The issue has been discussed on the `libtool` [mailing list](https://lists.gnu.org/archive/html/libtool/2021-06/msg00001.html).

**Comment by rmyorston** [comment]
Actually, the problem in `libtool` is a consequence of setting the build OS to `mingw64`. That gets _some_ things right for w64devkit but the filename conversion is wrong because we aren't using MSYS2. Other things may be wrong too.

We can only get so far by hacking GNU Autotools as they currently exist. The ultimate solution is to have proper support for the w64devkit environment in Autotools.

Anyway, here's a script that can be sourced before running `configure`:
```sh
export PATH_SEPARATOR=';'
export ac_executable_extensions='.exe'
export build_alias="$(uname -m)-pc-mingw64"

if [ -f configure ]
then
    echo "Converting configure..."
    sed -i 's/func_convert_file_msys_to_w32/func_convert_file_noop/' configure
fi
```
I've used this to successfully build [Expat](https://libexpat.github.io/) with w64devkit.

**Comment by ale5000-git** [comment]
After some reasearch I have found that instead of using:
`cmd //c echo`
they could achieve the same effect with:
`MSYS_NO_PATHCONV=1 cmd /c echo`

**Comment by rmyorston** [comment]
>MSYS_NO_PATHCONV=1 cmd /c echo

If setting that variable prevents MSYS-style paths being converted to DOS-style, then no, it wouldn't have the same effect. The whole point of the `cmd //c echo path` thing is to convert `path`.

**Comment by avih** [comment]
It prevents interpreting //c as /c. I.e. if autotools used this form, then they wouldn't have needed to use //c, and it would work in busybox-w32 out of the box.

But as it stands, they do use //c and expect msys2 env, so to run in busybox-w32 something needs to convert it to /c, and one method a recent commit in bb-w32 which does this conversion back to /c .

**Comment by ale5000-git** [comment]
> If setting that variable prevents MSYS-style paths being converted to DOS-style, then no, it wouldn't have the same effect. The whole point of the `cmd //c echo path` thing is to convert `path`.

Actually it is the opposite:
when executing external apps from bash if you run:
`cmd /c echo`
it become:
`cmd \c echo`
and so it doesn't work.

Their conversion is pretty lame.

Instead using:
`cmd //c echo`
or
`MSYS_NO_PATHCONV=1 cmd /c echo`
prevent the conversion and so it works.

**Comment by ale5000-git** [comment]
They even include a shell script named "start" in the PATH that contains:
`"$COMSPEC" //c start "${@//&/^&}"`
so they happily make it not working in any other apps once again.

**Comment by avih** [comment]
> When executing external apps from bash if you run, for example:
> `cmd /c echo`
> it become:
> `cmd \c echo`

No. `cmd /c` becomes `cmd c:\` .

**Comment by ale5000-git** [comment]
> No. `cmd /c` becomes `cmd c:\`

Yes, you are right; but the point is that this lame conversion clearly break almos everything.
